bundle/src/main/java/com/adobe/acs/commons/mcp/impl/processes/TreeReplication.java

```diff
@@ -0,0 +1,272 @@
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.mcp.impl.processes;
+
+import com.adobe.acs.commons.fam.ActionManager;
+import com.adobe.acs.commons.functions.CheckedFunction;
+import com.adobe.acs.commons.mcp.ProcessDefinition;
+import com.adobe.acs.commons.mcp.ProcessInstance;
+import com.adobe.acs.commons.mcp.form.CheckboxComponent;
+import com.adobe.acs.commons.mcp.form.FormField;
+import com.adobe.acs.commons.mcp.form.PathfieldComponent;
+import com.adobe.acs.commons.mcp.form.SelectComponent;
+import com.adobe.acs.commons.mcp.form.TextfieldComponent;
+import com.adobe.acs.commons.mcp.model.GenericReport;
+import com.adobe.acs.commons.util.visitors.TreeFilteringResourceVisitor;
+import com.day.cq.commons.jcr.JcrConstants;
+import com.day.cq.replication.AgentFilter;
+import com.day.cq.replication.ReplicationActionType;
+import com.day.cq.replication.ReplicationOptions;
+import com.day.cq.replication.Replicator;
+import com.day.cq.wcm.api.NameConstants;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.PersistenceException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.jcr.resource.api.JcrResourceConstants;
+
+/**
+ * Replace folder thumbnails under a user-definable set of circumstances As a
+ * business user, I would like an easy way to scan and repair missing
+ * thumbnails, or just regenerate all thumbnails under a given tree of the DAM.
+ */
+public class TreeReplication extends ProcessDefinition {
+
+    protected static enum ReplicationFilter {
+        ALL(r -> true),
+        FOLDERS_AND_PAGES_ONLY(TreeReplication::isFolderOrPage);
+
+        CheckedFunction<Resource, Boolean> test;
+
+        @SuppressWarnings("squid:UnusedPrivateMethod")
+        private ReplicationFilter(CheckedFunction<Resource, Boolean>... tests) {
+            this.test = CheckedFunction.or(tests);
+        }
+
+        @SuppressWarnings("squid:S00112")
+        public boolean shouldReplicate(Resource r) throws Exception {
+            return this.test.apply(r);
+        }
+    }
+
+    protected static enum QueueMethod {
+        USE_PUBLISH_QUEUE, USE_MCP_QUEUE, MCP_AFTER_10K
+    }
+
+    protected static enum ReplicationAction {
+        PUBLISH, UNPUBLISH
+    }
+
+    public static int ASYNC_LIMIT = 10000;
+
+    Replicator replicatorService;
+
+    @FormField(name = "Starting Path",
+            component = PathfieldComponent.FolderSelectComponent.class,
+            description = "This item and its descendants will be published/unpublished")
+    private String startingPath = "/content/dam";
```

Comment:
This is an odd default startingPath for the default of FOLDERS_AND_PAGES_ONLY? ..

Reply:
Oh that's true but the reason why its "Folders and Pages" is because pages have a duality of container + content -- there's really no way to say "Just these pages and not others" using the tree visitor -- it's node-type based.

I also didn't want it to default to /content because someone could wreak havoc not paying attention.  I could also make it blank by default if that makes more sense.  Actually, it might be blank in the UI because I didn't specify the default in the options of the FormField annotation.

Reply:
That's fair; I think ~everyone will be adjusting the path in some regard anyhow so not a big deal - just struck me as strange :)

Reply:
100% agreed.  Let's leave it blank so it's clear that the user has to specify it in the first place.